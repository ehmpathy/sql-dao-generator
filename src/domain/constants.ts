export enum DatabaseLanguage {
  MYSQL = 'mysql',
  POSTGRES = 'postgres',
}
