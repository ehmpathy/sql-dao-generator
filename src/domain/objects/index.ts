export * from './GeneratedCodeFile';
export * from './GeneratorConfig';
export * from './SqlSchemaPropertyMetadata';
export * from './SqlSchemaReferenceMetadata';
export * from './SqlSchemaToDomainObjectRelationship';
