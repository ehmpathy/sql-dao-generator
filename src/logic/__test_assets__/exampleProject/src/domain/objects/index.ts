export * from './AsyncTaskPredictStationCongestion'
export * from './Carriage';
export * from './Certificate';
export * from './Engineer';
export * from './Geocode';
export * from './Train';
export * from './Locomotive';
export * from './TrainLocatedEvent';
export * from './Station';
export * from './Invoice';
export * from './InvoiceLineItem';
export * from './Price';
