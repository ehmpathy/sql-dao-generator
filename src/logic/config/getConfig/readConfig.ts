import { DatabaseLanguage, GeneratorConfig } from '../../../domain';
import { readYmlFile } from '../../../utils/fileio/readYmlFile';
import { getDirOfPath } from '../../../utils/filepaths/getDirOfPath';
import { UserInputError } from '../../UserInputError';
import { extractDomainObjectMetadatasFromConfigCriteria } from './extractDomainObjectMetadatasFromConfigCriteria';
import { getAllPathsMatchingGlobs } from './getAllPathsMatchingGlobs';
import { getRelevantContentsOfSqlCodeGeneratorConfig } from './referencedConfigs/getRelevantContentsOfSqlCodeGeneratorConfig';
import { getRelevantContentsOfSqlSchemaControlConfig } from './referencedConfigs/getRelevantContentsOfSqlSchemaControlConfig';
import { getRelevantContentsOfSqlSchemaGeneratorConfig } from './referencedConfigs/getRelevantContentsOfSqlSchemaGeneratorConfig';

/*
  1. read the config
  2. validate the config
*/
export const readConfig = async ({
  configPath,
}: {
  configPath: string;
}): Promise<GeneratorConfig> => {
  const configDir = getDirOfPath(configPath);
  const getAbsolutePathFromRelativeToConfigPath = (relpath: string) =>
    `${configDir}/${relpath}`;

  // get the yml
  const contents = await readYmlFile({ filePath: configPath });

  // get the language and dialect
  if (!contents.language)
    throw new UserInputError({ reason: 'config.language must be defined' });
  const language = contents.language;
  if (contents.language && contents.language !== DatabaseLanguage.POSTGRES)
    throw new UserInputError({
      reason:
        'dao generator only supports postgres. please update the `language` option in your config to `postgres` to continue',
    });
  if (!contents.dialect)
    throw new UserInputError({ reason: 'config.dialect must be defined' });
  const dialect = `${contents.dialect}`; // ensure that we read it as a string, as it could be a number

  // validate the output config
  if (!contents.generates)
    throw new UserInputError({
      reason: 'config.generates key must be defined',
    });
  if (!contents.generates.daos?.to)
    throw new UserInputError({
      reason:
        'config.generates.daos.to must specify where to output the generated dao',
    });
  if (!contents.generates.daos?.using?.log)
    throw new UserInputError({
      reason:
        'config.generates.daos.using.log must specify where to find your log functions',
    });
  if (!contents.generates.daos?.using?.DatabaseConnection)
    throw new UserInputError({
      reason:
        'config.generates.daos.using.DatabaseConnection must specify where to find your DatabaseConnection type definition',
    });
  if (!contents.generates.schema?.config)
    throw new UserInputError({
      reason:
        'config.generates.schema must specify the path to your sql-schema-generator config',
    });
  if (!contents.generates.control?.config)
    throw new UserInputError({
      reason:
        'config.generates.control must specify the path to your sql-schema-control config',
    });
  if (!contents.generates.code?.config)
    throw new UserInputError({
      reason:
        'config.generates.code must specify the path to your sql-code-generator config',
    });
  const generates: GeneratorConfig['generates'] = {
    daos: {
      to: contents.generates.daos.to,
      using: {
        log: contents.generates.daos.using.log,
        DatabaseConnection: contents.generates.daos.using.DatabaseConnection,
      },
    },
    schema: {
      config: {
        path: contents.generates.schema.config,
        content: await getRelevantContentsOfSqlSchemaGeneratorConfig({
          pathToConfig: getAbsolutePathFromRelativeToConfigPath(
            contents.generates.schema.config,
          ),
        }),
      },
    },
    control: {
      config: {
        path: contents.generates.control.config,
        content: await getRelevantContentsOfSqlSchemaControlConfig({
          pathToConfig: getAbsolutePathFromRelativeToConfigPath(
            contents.generates.control.config,
          ),
        }),
      },
    },
    code: {
      config: {
        path: contents.generates.code.config,
        content: await getRelevantContentsOfSqlCodeGeneratorConfig({
          pathToConfig: getAbsolutePathFromRelativeToConfigPath(
            contents.generates.code.config,
          ),
        }),
      },
    },
  };

  // get the domain objects declarations
  if (!contents.for?.objects?.search)
    throw new UserInputError({
      reason:
        'config.for.objects.search must specify which files to start the search for domain objects [format: glob]',
    });
  const searchGlobs: string | string[] = contents.for.objects.search;
  // if (!contents.for?.objects?.include)
  //   console.log('config.for.objects.include was not defined, including all domain objects found by default');
  const include: string | string[] | null =
    contents.for?.objects?.include ?? null;
  // if (!contents.for?.objects?.exclude)
  //   console.log('config.for.objects.exclude was not defined, not excluding any domain objects found by default');
  const exclude: string | string[] | null =
    contents.for?.objects?.exclude ?? null;
  const relativeSearchPaths = await getAllPathsMatchingGlobs({
    globs: Array.isArray(searchGlobs) ? searchGlobs : [searchGlobs],
    root: configDir,
  });
  const domainObjectMetadatas =
    await extractDomainObjectMetadatasFromConfigCriteria({
      searchPaths: relativeSearchPaths.map((relPath) =>
        getAbsolutePathFromRelativeToConfigPath(relPath),
      ), // give absolute paths
      include: Array.isArray(include) ? include : include ? [include] : null,
      exclude: Array.isArray(exclude) ? exclude : exclude ? [exclude] : null,
    });

  // return the results
  return new GeneratorConfig({
    language,
    dialect,
    rootDir: configDir,
    for: {
      objects: domainObjectMetadatas,
    },
    generates,
  });
};
