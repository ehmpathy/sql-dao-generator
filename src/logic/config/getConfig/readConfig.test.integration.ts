import { GeneratorConfig } from '../../../domain';
import { readConfig } from './readConfig';

describe('readConfig', () => {
  it('should be able to read the example config provisioned in __test_assets__', async () => {
    const config = await readConfig({
      configPath: `${__dirname}/../../__test_assets__/exampleProject/codegen.sql.dao.yml`,
    });
    expect(config).toBeInstanceOf(GeneratorConfig);
    expect(config.language).toEqual('postgres');
    expect(config.dialect).toEqual('10.7');
    expect(config.generates).toEqual({
      daos: {
        to: 'src/data/dao',
        using: {
          log: 'src/util/log#log',
          DatabaseConnection: 'src/util/database/getDbConnection#DatabaseConnection',
        },
        schema: {
          config: 'codegen.schema.yml',
        },
        control: {
          config: 'provision/schema/control.yml',
        },
        code: {
          config: 'codegen.sql.yml',
        },
      },
    });
    expect(config.for.objects.length).toEqual(6);
    expect(config).toMatchSnapshot({ rootDir: expect.anything() }); // to log an example of the output; ignore the rootDir, to make it machine independent
  });
});
