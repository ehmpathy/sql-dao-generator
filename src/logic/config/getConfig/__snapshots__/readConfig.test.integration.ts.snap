// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readConfig should be able to read the example config provisioned in __test_assets__ 1`] = `
Object {
  "dialect": "10.7",
  "for": Object {
    "objects": Array [
      DomainObjectMetadata {
        "decorations": Object {
          "unique": Array [
            "cin",
          ],
          "updatable": Array [],
        },
        "extends": "DomainEntity",
        "name": "Carriage",
        "properties": Object {
          "capacity": DomainObjectPropertyMetadata {
            "name": "capacity",
            "nullable": false,
            "required": true,
            "type": "NUMBER",
          },
          "carries": DomainObjectPropertyMetadata {
            "name": "carries",
            "nullable": false,
            "of": Array [
              "PASSENGER",
              "FREIGHT",
            ],
            "required": true,
            "type": "ENUM",
          },
          "cin": DomainObjectPropertyMetadata {
            "name": "cin",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "id": DomainObjectPropertyMetadata {
            "name": "id",
            "nullable": false,
            "required": false,
            "type": "NUMBER",
          },
          "uuid": DomainObjectPropertyMetadata {
            "name": "uuid",
            "nullable": false,
            "required": false,
            "type": "STRING",
          },
        },
      },
      DomainObjectMetadata {
        "decorations": Object {
          "unique": null,
          "updatable": null,
        },
        "extends": "DomainValueObject",
        "name": "Geocode",
        "properties": Object {
          "id": DomainObjectPropertyMetadata {
            "name": "id",
            "nullable": false,
            "required": false,
            "type": "NUMBER",
          },
          "latitude": DomainObjectPropertyMetadata {
            "name": "latitude",
            "nullable": false,
            "required": true,
            "type": "NUMBER",
          },
          "longitude": DomainObjectPropertyMetadata {
            "name": "longitude",
            "nullable": false,
            "required": true,
            "type": "NUMBER",
          },
        },
      },
      DomainObjectMetadata {
        "decorations": Object {
          "unique": Array [
            "ein",
          ],
          "updatable": Array [
            "milage",
          ],
        },
        "extends": "DomainEntity",
        "name": "Locomotive",
        "properties": Object {
          "capacity": DomainObjectPropertyMetadata {
            "name": "capacity",
            "nullable": false,
            "required": true,
            "type": "NUMBER",
          },
          "ein": DomainObjectPropertyMetadata {
            "name": "ein",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "fuel": DomainObjectPropertyMetadata {
            "name": "fuel",
            "nullable": false,
            "of": Array [
              "STEAM",
              "COAL",
              "DIESEL",
              "BATTERY",
              "FISSION",
            ],
            "required": true,
            "type": "ENUM",
          },
          "id": DomainObjectPropertyMetadata {
            "name": "id",
            "nullable": false,
            "required": false,
            "type": "NUMBER",
          },
          "milage": DomainObjectPropertyMetadata {
            "name": "milage",
            "nullable": false,
            "required": true,
            "type": "NUMBER",
          },
          "uuid": DomainObjectPropertyMetadata {
            "name": "uuid",
            "nullable": false,
            "required": false,
            "type": "STRING",
          },
        },
      },
      DomainObjectMetadata {
        "decorations": Object {
          "unique": Array [
            "combinationId",
          ],
          "updatable": Array [
            "engineers",
            "status",
          ],
        },
        "extends": "DomainEntity",
        "name": "Train",
        "properties": Object {
          "carriageUuids": DomainObjectPropertyMetadata {
            "name": "carriageUuids",
            "nullable": false,
            "of": Object {
              "nullable": false,
              "required": true,
              "type": "STRING",
            },
            "required": true,
            "type": "ARRAY",
          },
          "combinationId": DomainObjectPropertyMetadata {
            "name": "combinationId",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "engineerUuids": DomainObjectPropertyMetadata {
            "name": "engineerUuids",
            "nullable": false,
            "of": Object {
              "nullable": false,
              "required": true,
              "type": "STRING",
            },
            "required": true,
            "type": "ARRAY",
          },
          "homeStationGeocode": DomainObjectPropertyMetadata {
            "name": "homeStationGeocode",
            "nullable": false,
            "of": DomainObjectReferenceMetadata {
              "extends": "DomainValueObject",
              "name": "Geocode",
            },
            "required": true,
            "type": "REFERENCE",
          },
          "id": DomainObjectPropertyMetadata {
            "name": "id",
            "nullable": false,
            "required": false,
            "type": "NUMBER",
          },
          "leadEngineerUuid": DomainObjectPropertyMetadata {
            "name": "leadEngineerUuid",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "locomotiveUuids": DomainObjectPropertyMetadata {
            "name": "locomotiveUuids",
            "nullable": false,
            "of": Object {
              "nullable": false,
              "required": true,
              "type": "STRING",
            },
            "required": true,
            "type": "ARRAY",
          },
          "status": DomainObjectPropertyMetadata {
            "name": "status",
            "nullable": false,
            "of": Array [
              "ASSEMBLED",
              "DISASSEMBLED",
            ],
            "required": true,
            "type": "ENUM",
          },
          "uuid": DomainObjectPropertyMetadata {
            "name": "uuid",
            "nullable": false,
            "required": false,
            "type": "STRING",
          },
        },
      },
      DomainObjectMetadata {
        "decorations": Object {
          "unique": Array [
            "socialSecurityNumberHash",
          ],
          "updatable": Array [
            "name",
          ],
        },
        "extends": "DomainEntity",
        "name": "TrainEngineer",
        "properties": Object {
          "id": DomainObjectPropertyMetadata {
            "name": "id",
            "nullable": false,
            "required": false,
            "type": "NUMBER",
          },
          "name": DomainObjectPropertyMetadata {
            "name": "name",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "socialSecurityNumberHash": DomainObjectPropertyMetadata {
            "name": "socialSecurityNumberHash",
            "nullable": false,
            "required": true,
            "type": "STRING",
          },
          "uuid": DomainObjectPropertyMetadata {
            "name": "uuid",
            "nullable": false,
            "required": false,
            "type": "STRING",
          },
        },
      },
    ],
  },
  "generates": Object {
    "code": Object {
      "config": Object {
        "content": Object {
          "generates": Object {
            "queryFunctions": "src/data/dao/.generated/queryFunctions.ts",
            "types": "src/data/dao/.generated/types.ts",
          },
          "queries": Array [
            "src/data/dao/**/*.ts",
            "!src/data/dao/**/index.ts",
            "!src/data/dao/**/castFromDatabaseObject.ts",
            "!src/**/*.test.ts",
          ],
          "resources": Array [
            "provision/schema/sql/tables/**/*.sql",
            "provision/schema/sql/views/**/*.sql",
            "provision/schema/sql/functions/**/*.sql",
          ],
        },
        "path": "codegen.sql.types.yml",
      },
    },
    "control": Object {
      "config": Object {
        "content": Object {
          "definitions": Array [
            "./sql/domain.yml",
          ],
        },
        "path": "provision/schema/control.yml",
      },
    },
    "daos": Object {
      "to": "src/data/dao",
      "using": Object {
        "DatabaseConnection": "src/util/database/getDbConnection#DatabaseConnection",
        "log": "src/util/log#log",
      },
    },
    "schema": Object {
      "config": Object {
        "content": Object {
          "declarations": "provision/schema/declarations/index.ts",
          "generates": Object {
            "sql": Object {
              "to": "provision/schema/sql",
            },
          },
        },
        "path": "codegen.sql.schema.yml",
      },
    },
  },
  "language": "postgres",
  "rootDir": Anything,
}
`;
